waiter/src/waiter/shell_scheduler.clj

```diff
@@ -128,10 +128,11 @@
 
 (defn launch-instance
   "Launches a new process for the given service-id"
-  [service-id working-dir-base-path command environment port->reservation-atom port-range]
+  [service-id working-dir-base-path command environment num-ports port->reservation-atom port-range]
   (when-not command
     (throw (ex-info "The command to run was not supplied" {:service-id service-id})))
-  (let [port (reserve-port! port->reservation-atom port-range)
+  (let [reserved-ports (repeatedly num-ports #(reserve-port! port->reservation-atom port-range))
```

Comment:
The port count should be passed to `reserve-port!`.  What happens if you run out of ports?